src/commands/watch.rs

```diff
@@ -185,48 +175,49 @@ impl WatchCommand {
             .await?
             .into_v1();
 
-        let path_patterns: Vec<String> = match self.watch_assets {
-            // Watch patterns
-            false => app_manifest
-                .components
-                .iter()
-                .filter_map(|c| c.build.as_ref())
-                .filter_map(|b| b.watch.clone())
-                .flatten()
+        let watch_patterns = app_manifest
+            .components
+            .iter()
+            .filter_map(|c| c.build.as_ref())
+            .filter_map(|b| b.watch.clone())
+            .flatten();
+        let component_source_patterns = app_manifest
+            .components
+            .iter()
+            .filter_map(|c| {
+                if let RawModuleSource::FileReference(path) = &c.source {
+                    return path.to_str();
+                }
+                None
+            })
+            .map(String::from);
+        let component_file_patterns = app_manifest
+            .components
+            .iter()
+            .filter_map(|c| c.wasm.files.as_ref())
+            .flatten()
+            .map(|raw_file_mount| match raw_file_mount {
+                RawFileMount::Placement(raw_directory_placement) => String::from(
+                    raw_directory_placement
+                        .source
+                        .join("**/*")
+                        .to_str()
+                        .expect("conversion to str not to fail"),
+                ),
+                RawFileMount::Pattern(pattern) => pattern.to_string(),
+            });
+        let path_patterns: Vec<String> = match self.skip_build {
+            false => watch_patterns.chain(component_file_patterns).collect(),
```

Comment:
~~The more I thought about this yesterday the less sure I was about doing this. For example, what do we do for a spin app with two components: one where this is a build command and one where there isn't. Should we skip the build step and watch for changes to wasm because they'll have to manually build one part of the app? Or should we run the build step because at least one component needs it and forgo watching the wasm.~~

I misheard you and was thinking about `build.command` not `build.watch`.

I personally think that if they run `spin watch` with no `build.watch` we should:
1. Inform them they haven't configured `build.watch`
2. Point them to docs on how to do it
3. Or ask them to run `spin watch --skip-build` if they care to manually build themselves.

My take right now is that we should just assume that users with no `build.watch` entry

Reply:
Hmm, if I have an app with a Rust component (has a `build`, has a `build.watch`) and a fileserver (no `build.watch`), and I `spin watch` that, does this:

1. Reject because the fileserver doesn't have a `build.watch` and ask me to configure one; or
2. Accept because the Rust component has a `build.watch`, and ignore if I manually update the fileserver Wasm; or
3. Something else?

I am trying to think what the "it just works" behaviour would be in this situation...!

Reply:
Maybe having a `build.command` but no `build.watch` is the 'did you forget' case (no `build.command` = just watch the Wasm).

Reply:
Hmm good point. So I guess it would do it on a per component basis.

When running `spin watch` for each component:
- If it has a `build.command` and a `build.watch` -> watch `build.watch` and `component.files`
- If it has a `build.command` and no `build.watch` -> throw a warning and watch `component.files`
- If it has no `build.command` and a `build.watch` -> watch `build.watch` and `component.source` and `component.files`
- If i has no `build.command` and no `build.watch` -> watch `component.source` and `component.files`


When running `spin watch --skip-build` for each component:
- We will just always watch `component.source` and `component.files`

How's that?

Reply:
That sounds good!  My only suggested tweak would be in case 2 I would watch `component.source` anyway, but given the warning I don't feel too strongly about this.

Nice catch on option 3.  I don't _think_ that should be possible (the schema should fail to validate if `build` exists but has no `command`) but yes to belt and braces!

Reply:
If we watch `component.source` in case 2 we get a circular dependency and infinite rebuilding.

Reply:
Oh, right, because we currently treat all changes as requiring a rebuild.  Yep, as you were.  (But that's another reason to put that on the to-do list...!  _grin_)